Update to the latest version of MK

hellais opened the pull request:
go-measurement-kit PR related to https://github.com/measurement-kit/script-build-unix/pull/8.

Still haven't tested that everything works properly with this new version, so this is mostly about making the tooling around it a little bit better.

I am also missing a prebuilt version for linux. It would be useful if somebody with a linux box handy could check that out (cc @sarathms).

hellais committed:
Update the download-libs.sh script to a new pattern

hellais committed:
Quick and dirty script to generate the golang include header

hellais committed:
Bump MK versions in go-measurement-kit

bassosimone commented on download-libs.sh:
```diff
@@ -1,28 +1,47 @@
+7c62fae040b242ffa3de5fc4635de69ac440af2b4401b016fc0e197c7abe3962  mingw-libressl-2.8.3+1.tar.gz
+8cad84a3253b14377384148163eccf1c26765a65d3c4487d47726b9a08a9bbc7  mingw-measurement-kit-0.9.1+1.tar.gz
+END
+)
```
You _could_ use script-build-unix as a sub repository and its `./install` script instead.

hellais replied:
Sub-repositories don't play nicely with `go dep` so I can't do that unfortunately :(

bassosimone commented:
ah, right!

hellais committed:
Add a very basic unittest

hellais committed:
Bump travis

hellais commented:
@bassosimone do you have pointers as to what I should be looking out for which has changed between v0.9.0 and v0.9.1?

For example I got now the error:
```
# github.com/measurement-kit/go-measurement-kit
./measurement_kit.go:171:10: could not determine kind of name for C.mk_task_start
```

bassosimone commented:
Use C.mk_nettest_start (or mk_start_nettest don’t remember by hearth which one is true but I believe the former is correct).

I think no other API changed.

hellais committed:
Update to latest version of the MK API

hellais committed:
Add a cert.pem for testing

hellais committed:
Better tooling around testing

hellais committed:
Make the unittests a little bit more useful

hellais commented on task_data.go:
```diff
@@ -27,12 +20,12 @@ type TaskData struct {
+	SaveRealProbeCC  bool    `json:"save_real_probe_cc,omitempty"`
+	NoCollector      bool    `json:"no_collector,omitempty"`
+	NoFileReport     bool    `json:"no_file_report,omitempty"`
+	RandomizeInput   bool    `json:"randomize_input,omitempty"`
```
This was the other API change

bassosimone replied:
Right, thanks!

hellais committed:
Add support for linux

Drop misleading arm support

hellais committed:
Add linux checks to travis

hellais commented:
@bassosimone the linux travis build is failing due to:

```
./libs/MK_DIST/linux/measurement-kit/0.9.1+1/x86_64/lib/libmeasurement_kit.a(ffi.o): In function `void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) [clone .isra.78]':
ffi.cpp:(.text+0x109): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_create(unsigned long&, unsigned long)'
ffi.cpp:(.text+0x13f): undefined reference to `std::__throw_logic_error(char const*)'
./libs/MK_DIST/linux/measurement-kit/0.9.1+1/x86_64/lib/libmeasurement_kit.a(ffi.o): In function `mk_event_destroy':
```

I suspect this has to do with some C++11 ABI incompatibilities in the linux static build. See: https://github.com/measurement-kit/script-build-unix/pull/10

Do you have ideas on how I should be building the linux version of MK to avoid these linking issues on linux?
Is this a lost battle because we will always find some version of linux which is using an ABI incompatible C++ library?

bassosimone commented:
@hellais not necessarily, but I suspect the default linux environment could be too old. Is it possible for you to upgrade the build to use a more modern version of Ubuntu?

hellais committed:
Try using xenial instead of trusty

hellais commented:
> not necessarily, but I suspect the default linux environment could be too old. Is it possible for you to upgrade the build to use a more modern version of Ubuntu?

I get the same error when using xenial (the build was done on bionic) and the default build env was trusty, which is actually not that old.

hellais committed:
Order matters

hellais committed:
Re-ordering more

hellais committed:
Take lib order from: https://github.com/measurement-kit/android-libs/blob/master/jni/Android.mk.in#L53

hellais committed:
Add a dummy.cc file

hellais commented:
After a fair amount of testing and a lot of help from @bassosimone, we came to the bottom of this.

It seems like there were two separate problems at hand:

1.  The golang tooling was not recognising that it should be using a C++ compiler and therefore was not linking to the C++11 standard library. We were already forcing this via some `LDFLAGS`, but that wasn't working as expected on linux.

It seems that just dropping a `dummy.cc` file inside of the root of the golang repo tree, tell the golang cgo compiler to use `g++` and everything works much better.

2. The order in which the static libraries are linked in matters and I had jumbled the order since using a script to auto-generate them.
The correct library linking order is: `libmeasurement_kit.a libmaxminddb.a libcurl.a libevent_openssl.a libssl.a libcrypto.a libevent_core.a libevent_extra.a libevent_pthreads`

I am going to work a bit on cleaning this up following these two conclusions, namely:

* [x] Dropping the now unncessary `-lstdc++` `LDFLAGS` from the `cgo` directives
* [x] Adjusting the header generation script to sort the libraries properly

hellais committed:
Drop C++ related LDFLAGS

hellais committed:
Update SHASUMS for linux build

hellais committed:
Update gen-include-header script

hellais committed:
Update generated header

hellais committed:
Add go build command to travis

hellais committed:
Fix gen-include-header script

hellais committed:
Update cgo directives

hellais committed:
Try re-adding -lstdc++

hellais commented:
Ugh, this is still failing on `xenial` :(: https://travis-ci.com/measurement-kit/go-measurement-kit/jobs/171022940.

Yet the docker based testing harness that runs on ubuntu `bionic` works fine on bionic. See: https://github.com/measurement-kit/script-build-unix/pull/10.

I think more debugging is necessary and probably a better solution is to make go-measurement-kit on linux not rely on pre-built packages upload to `script-build-linux`, but rather the linux user should build their own measurement-kit static library and link it themselves!

bassosimone commented:
I have changed the title of this PR because we want to use the latest version of MK that is available at the moment, rather than using a version that at this point is old (i.e. `v0.9.1`).

bassosimone committed:
Make this repository work well with macOS

```
$ otool -L ./ndt
./ndt:
	/usr/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
```

I think we're good here.

Xref: https://github.com/measurement-kit/script-build-unix/issues/15#issuecomment-490893951

bassosimone committed:
Better readme instructions

bassosimone committed:
Make sure we can build static binaries on Linux

This is based on https://github.com/measurement-kit/script-build-linux.

I like this solution based on Docker as it's really simple. Starting from
the Android build script, I did this in two hours.

bassosimone committed:
Make working for Windows from macOS pleasant

bassosimone committed:
Start removing unnecessary stuff

bassosimone committed:
Merge remote-tracking branch 'origin/master' into update/mk-0.9.1

bassosimone committed:
.travis.yml: yak shaving

bassosimone committed:
Rewrite the testing strategy

bassosimone committed:
x

bassosimone committed:
clarify

bassosimone commented on measurement_kit.go:
```diff
@@ -3,45 +3,40 @@ package mk
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/libressl/2.6.4-4/armv7/lib/libcrypto.a
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/libevent/2.1.8-4/armv7/lib/libevent.a
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/libevent/2.1.8-4/armv7/lib/libevent_openssl.a
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/libevent/2.1.8-4/armv7/lib/libevent_pthreads.a
```
Note: I have deliberately removed support for `linux,arm`. We don't want more bloodshed. We actually want to have another reason to have as much as possible of MK in Go to recompile in three minutes.

bassosimone replied:
Yet, if we really want that, the procedure is straightforward. Just create antother Dockerfile like the one used in script-build-linux, except that armv7 is used. (But, do we really want to invest time in suffering where we could invest time in suffering less soon-ish?)

bassosimone committed:
this changed again

bassosimone committed:
Make the build work locally

bassosimone commented:
All green. I feel entitled to merge this PR. I'm meeting with @hellais next week anyway to move forward the OONI Desktop implementation. I think a reasonable prototype for then it's useful.
